internal/pkg/initworkload/workload.go

```diff
@@ -0,0 +1,325 @@
+
+package initworkload
+
+import (
+	"encoding"
+	"errors"
+	"fmt"
+	"path/filepath"
+
+	"github.com/aws/copilot-cli/internal/pkg/config"
+	"github.com/aws/copilot-cli/internal/pkg/manifest"
+	"github.com/aws/copilot-cli/internal/pkg/term/color"
+	"github.com/aws/copilot-cli/internal/pkg/term/log"
+	"github.com/aws/copilot-cli/internal/pkg/workspace"
+)
+
+const (
+	fmtAddWlToAppStart    = "Creating ECR repositories for %s %s."
+	fmtAddWlToAppFailed   = "Failed to create ECR repositories for %s %s.\n"
+	fmtAddWlToAppComplete = "Created ECR repositories for %s %s.\n"
+)
+
+const (
+	jobWlType = "job"
+	svcWlType = "service"
+)
+
+// Store represents the methods needed to add workloads to the SSM parameter store.
+type Store interface {
+	GetApplication(appName string) (*config.Application, error)
+	CreateService(service *config.Workload) error
+	CreateJob(job *config.Workload) error
+	ListServices(appName string) ([]*config.Workload, error)
+	ListJobs(appName string) ([]*config.Workload, error)
+}
+
+// WorkloadAdder contains the methods needed to add jobs and services to an existing application.
+type WorkloadAdder interface {
+	AddJobToApp(app *config.Application, jobName string) error
+	AddServiceToApp(app *config.Application, serviceName string) error
+}
+
+// Workspace contains the methods needed to manipulate a Copilot workspace.
+type Workspace interface {
+	CopilotDirPath() (string, error)
+	WriteJobManifest(marshaler encoding.BinaryMarshaler, jobName string) (string, error)
+	WriteServiceManifest(marshaler encoding.BinaryMarshaler, serviceName string) (string, error)
+}
+
+// Prog contains the methods needed to render multi-stage operations.
+type Prog interface {
+	Start(label string)
+	Stop(label string)
+}
+
+// WorkloadProps contains the information needed to represent a Workload (job or service)
+type WorkloadProps struct {
+	App            string
+	Type           string
+	Name           string
+	DockerfilePath string
+	Image          string
+
+	Schedule string
+	Timeout  string
+	Retries  int
+
+	Port        uint16
+	HealthCheck *manifest.ContainerHealthCheck
+}
+
+func (p *WorkloadProps) isValidWorkload() bool {
+	return p.App != "" &&
+		p.Type != "" &&
+		p.Name != "" &&
+		(p.DockerfilePath != "" || p.Image != "")
+}
+
+func (p *WorkloadProps) isValidService() bool {
+	return p.isValidWorkload() &&
+		p.Schedule == "" &&
+		p.Timeout == "" &&
+		p.Retries == 0
+}
+
+func (p *WorkloadProps) isValidJob() bool {
+	return p.isValidWorkload() &&
+		p.Schedule != "" &&
+		p.Port == 0
+}
+
+type manifestWriter func(encoding.BinaryMarshaler, string) (string, error)
+
+type manifestCreator func(*WorkloadProps, Store) (encoding.BinaryMarshaler, error)
+
+type workloadAdder func(*config.Application, string) error
+
+type storeWorkloadAdder func(*config.Workload) error
+
+// WorkloadInitializer holds the clients necessary to initialize either a
+// service or job in an existing application.
+type WorkloadInitializer struct {
```

Comment:
Same with this struct. We could include only workload related (shared by service and job) methods in this struct. Then have `workloadInitializer`, `ServiceInitializer` and `JobInitializer`, and make those constructors return them respectively.

Reply:
There's not really enough separation between "workloadInitializer" and "ServiceInitializer" for me to want to do this just yet. Perhaps if we add mroe functionality to the package it will be worth separating the structs. But for the time being I've added the ServiceProps and JobProps structs for the inputs to Service() and Job() for better logical separation.